File: 2015/2015PracticeBot/src/Subsystems/SensorSubsystem.cpp
#include "SensorSubsystem.h"
#include "../RobotMap.h"
#include "../Commands/SensorUpdateCommand.h"
#include "../INS/INS.h"

const uint8_t IMU_UPDATE_RATE_HZ = 50;
const float FORCE_SLOPE = 83.568;
const float FORCE_Y_INTERCEPT = -24.404;

const float ULTRASONIC_SENSOR_SCALE = 113.0291799396;
const float ULTRASONIC_SENSOR_SCALE_MB1013 = 40.3149606299;

const float LOWER_CAR_MANIPULATOR_COUNTS_PER_INCH = 730.0;
const float LOWER_CAR_MANIPULATOR_OPEN_DISTANCE_WHEN_CLOSED = 12.375;

const float LOWER_TOWER_COUNTS_PER_INCH = 40.425355545341415285296475896619;

const float DRIVE_INCH_PER_COUNT = 0.1570796327;

const float PI = 3.14159265359;

SensorSubsystem::SensorSubsystem() : Subsystem("SensorSubsystem") {
	this->powerDistributionPanel = new PowerDistributionPanel();

	this->serialPort = new SerialPort(57600, SerialPort::kUSB);
	this->imu = new IMU(serialPort, IMU_UPDATE_RATE_HZ);
	if (imu) {
		LiveWindow::GetInstance()->AddSensor("IMU", "Gyro", imu);
	}

	this->leftDriveEncoder = new Encoder(0, 1 , false, Encoder::k4X);
	this->leftDriveEncoder->SetPIDSourceParameter(Encoder::kRate);
	this->leftDriveEncoder->SetDistancePerPulse(DRIVE_INCH_PER_COUNT);
	this->leftDriveEncoder->SetMaxPeriod(1.0);
	this->leftDriveEncoder->Reset();

	this->rightDriveEncoder = new Encoder(2, 3 , false, Encoder::k4X);
	this->rightDriveEncoder->SetPIDSourceParameter(Encoder::kRate);
	this->rightDriveEncoder->SetDistancePerPulse(DRIVE_INCH_PER_COUNT);
	this->rightDriveEncoder->SetMaxPeriod(1.0);
	this->rightDriveEncoder->Reset();

	this->centerDriveEncoder = new Encoder(4, 5 , false, Encoder::k4X);
	this->centerDriveEncoder->SetPIDSourceParameter(Encoder::kDistance);
	this->centerDriveEncoder->SetDistancePerPulse(DRIVE_INCH_PER_COUNT);
	this->centerDriveEncoder->SetMaxPeriod(1.0);
	this->centerDriveEncoder->Reset();


	INS::GetInstance()->Enable();
	INS::GetInstance()->Reset();

}

void SensorSubsystem::InitDefaultCommand() {
	SetDefaultCommand(new SensorUpdateCommand());
}

// Put methods for controlling this subsystem
// here. Call these from Commands.


IMU *SensorSubsystem::GetIMU() {
	return this->imu;
}

double SensorSubsystem::GetChannelCurrent(uint8_t channel)
{
	return this->powerDistributionPanel->GetCurrent(channel);
}

float SensorSubsystem::GetIMUYawRad() {
	return this->imu->GetYaw() * PI / 180.0;
}

Encoder *SensorSubsystem::GetLeftDriveEncoder() {
	return this->leftDriveEncoder;
}

Encoder *SensorSubsystem::GetRightDriveEncoder() {
	return this->rightDriveEncoder;
}

Encoder *SensorSubsystem::GetCenterDriveEncoder() {
	return this->centerDriveEncoder;
}

File: 2016/DriveBot/src/com/kennedyrobotics/robot2016/subsystems/SensorSubsystem.java
package com.kennedyrobotics.robot2016.subsystems;

import com.kauailabs.navx.frc.AHRS;
import com.kennedyrobotics.robot2016.RobotMap;
import com.kennedyrobotics.robot2016.commands.SensorUpdateCommand;

import edu.wpi.first.wpilibj.CounterBase.EncodingType;
import edu.wpi.first.wpilibj.DigitalInput;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.Encoder;
import edu.wpi.first.wpilibj.SerialPort;
import edu.wpi.first.wpilibj.command.Subsystem;
import edu.wpi.first.wpilibj.livewindow.LiveWindow;

/**
 *
 */
public class SensorSubsystem extends Subsystem {
    
	public static final double DRIVE_INCH_PER_COUNT = 0.1570796327;
	
    // Put methods for controlling this subsystem
    // here. Call these from Commands.

	private final Encoder chassisLeftEncoder = new Encoder(
			RobotMap.CHASSIS_ENCODER_LEFT_A, 
			RobotMap.CHASSIS_ENCODER_LEFT_B, 
			false, 
			EncodingType.k4X);
	private final Encoder chassisRightEncoder = new Encoder(
			RobotMap.CHASSIS_ENCODER_RIGHT_A, 
			RobotMap.CHASSIS_ENCODER_RIGHT_B, 
			false, 
			EncodingType.k4X);
	private final DigitalInput aimArmTiltLowLimit = new DigitalInput(RobotMap.ARM_TILT_LOW_LIMIT);
	private final DigitalInput aimArmTiltHighLimit = new DigitalInput(RobotMap.ARM_TILT_HIGH_LIMIT);
	private final DigitalInput shooterAimTiltLowLimit = new DigitalInput(RobotMap.BALL_SHOOTER_TILT_LOW_LIMIT);
	private final DigitalInput shooterAimTiltHighLimit = new DigitalInput(RobotMap.BALL_SHOOTER_TILT_HIGH_LIMIT);
	
	private AHRS navX;
	private boolean collision = false;
		
	public SensorSubsystem() {
		chassisLeftEncoder.setDistancePerPulse(DRIVE_INCH_PER_COUNT);
		chassisLeftEncoder.setMaxPeriod(1.0);
		chassisLeftEncoder.reset();
		
		chassisRightEncoder.setDistancePerPulse(DRIVE_INCH_PER_COUNT);
		chassisRightEncoder.setMaxPeriod(1.0);
		chassisRightEncoder.reset();
		
		try {
			navX = new AHRS(SerialPort.Port.kMXP);
		} catch (Exception e) {
			DriverStation.reportError("Error instantiating navX MXP:  " + e.getMessage(), true);
			navX = null;
		}
		if(navX != null) {
			LiveWindow.addSensor("IMU", "Gyro", navX);
		}
	}
	
    public void initDefaultCommand() {
    	setDefaultCommand(new SensorUpdateCommand());
    }
    
    public boolean getArmAimTiltLowLimit(){
    	return aimArmTiltLowLimit.get();
    }
    
    public boolean getArmAimTiltHighLimit(){
    	return aimArmTiltHighLimit.get();
    }
    
    public boolean getShooterAimTiltHighLimit(){
    	return shooterAimTiltHighLimit.get();
    }
    
    public boolean getShooterAimTiltLowLimit(){
    	return shooterAimTiltLowLimit.get();
    }
    
	public Encoder getChassisLeftEncoder() {
		return chassisLeftEncoder;
	}
	
	public Encoder getChassisRightEncoder() {
		return chassisRightEncoder;
	}

    
    public AHRS getNavX() {
    	return navX;
    }
    
	public double getYaw() {
		if(navX == null) {
			DriverStation.reportError("NavX is null", false);
			return 0.0;
		} else {
			return navX.getYaw();

		}
	}
	
	public void setCollision(boolean collision){
		this.collision = collision;
	}
	
	public boolean getCollision(){
		return collision;
	}
}

